File: python-server/modules/access_control.py
"""
Access control: agent validation and eligibility.
Mirrors accessControl.ts exactly.
"""
from datetime import datetime


def _get_week(date_str):
    """Returns the ISO week string for a date, e.g. '2025-W15'."""
    d = datetime.strptime(date_str, '%Y-%m-%d')
    # Match the JS logic exactly: use UTC-based ISO week calculation
    # JS: date.setUTCDate(date.getUTCDate() + 4 - dayOfWeek)
    day_of_week = d.isoweekday()  # Mon=1..Sun=7
    # In JS: getUTCDay() returns 0=Sun..6=Sat, then || 7 makes Sun=7
    # So dayOfWeek in JS is same as isoweekday() in Python
    import datetime as dt
    d2 = dt.date(d.year, d.month, d.day)
    iso_year, iso_week, _ = d2.isocalendar()
    return f"{iso_year}-W{str(iso_week).zfill(2)}"


def validate_agent_pickup(slots, slot_id, agent_id, roster):
    """Validates whether an agent can pick up a specific slot."""
    slot = next((s for s in slots if s['id'] == slot_id), None)
    if not slot:
        return {'valid': False, 'error': 'Slot not found.'}

    if slot['status'] != 'Released':
        return {'valid': False, 'error': 'This slot is no longer available.'}

    # 30-minute cutoff
    slot_start_time = slot['timeWindow'].split('-')[0].strip() if '-' in slot['timeWindow'] else None
    if slot_start_time:
        now = datetime.now()
        try:
            slot_dt = datetime.strptime(f"{slot['date']}T{slot_start_time}:00", '%Y-%m-%dT%H:%M:%S')
            diff_ms = (slot_dt - now).total_seconds() * 1000
            diff_minutes = diff_ms / (1000 * 60)
            if diff_minutes < 30:
                return {'valid': False, 'error': 'This slot can no longer be picked up (less than 30 minutes to start time).'}
        except ValueError:
            pass

    # Find agent's entries
    entries = roster.get('entries', [])
    agent_entries = [e for e in entries if e['agent'] == agent_id]
    if not agent_entries:
        return {'valid': False, 'error': 'Agent not found in roster.'}

    # Check program match
    agent_programs = list(set(e['program'] for e in agent_entries))
    if slot['program'] not in agent_programs:
        return {'valid': False, 'error': 'You are not eligible for this slot (program mismatch).'}

    # Check lobby match
    if slot.get('lobby'):
        agent_lobbies = list(set(e['lobby'] for e in agent_entries if e.get('lobby')))
        if agent_lobbies and slot['lobby'] not in agent_lobbies:
            return {'valid': False, 'error': 'You are not eligible for this slot (lobby mismatch).'}

    is_full_day = slot['otType'] == 'Full Day OT'

    if is_full_day:
        # Agent must have weekly off on that date
        date_entry = next((e for e in agent_entries if e['date'] == slot['date']), None)
        if not date_entry or not date_entry.get('isWeeklyOff'):
            return {'valid': False, 'error': 'Full Day OT is only available for agents with a weekly off on this date.'}

        # Max 1 Full Day OT per week when agent has 2+ weekly offs
        slot_week = _get_week(slot['date'])
        agent_wo_dates = [e['date'] for e in agent_entries if e.get('isWeeklyOff')]
        wo_in_week = len([d for d in agent_wo_dates if _get_week(d) == slot_week])

        if wo_in_week >= 2:
            filled_full_day_this_week = [
                s for s in slots
                if s['status'] == 'Filled'
                and s.get('filledByAgentId') == agent_id
                and s['otType'] == 'Full Day OT'
                and _get_week(s['date']) == slot_week
            ]
            if len(filled_full_day_this_week) >= 1:
                return {'valid': False, 'error': 'You have reached the maximum Full Day OT pickups for this week.'}
    else:
        # Pre/Post OT: slot's assigned agent must match requesting agent
        if slot.get('assignedAgentId') and slot['assignedAgentId'] != agent_id:
            return {'valid': False, 'error': 'This slot is assigned to a different agent.'}

        # Max 1 Pre/Post OT per agent per day
        filled_pre_post_today = [
            s for s in slots
            if s['status'] == 'Filled'
            and s.get('filledByAgentId') == agent_id
            and s['date'] == slot['date']
            and s['otType'] != 'Full Day OT'
        ]
        if len(filled_pre_post_today) >= 1:
            return {'valid': False, 'error': 'You already have a Pre/Post Shift OT pickup on this date.'}

    return {'valid': True}


def get_eligible_slots_for_agent(slots, agent_id, roster):
    """Returns eligible slots for an agent."""
    entries = roster.get('entries', [])
    agent_entries = [e for e in entries if e['agent'] == agent_id]
    if not agent_entries:
        return []

    agent_programs = set(e['program'] for e in agent_entries)
    agent_lobbies = set(e['lobby'] for e in agent_entries if e.get('lobby'))
    released_slots = [s for s in slots if s['status'] == 'Released']

    result = []
    for slot in released_slots:
        # Must match program
        if slot['program'] not in agent_programs:
            continue

        # Must match lobby
        if slot.get('lobby') and agent_lobbies and slot['lobby'] not in agent_lobbies:
            continue

        # 30-minute cutoff
        slot_start_time = slot['timeWindow'].split('-')[0].strip() if '-' in slot['timeWindow'] else None
        if slot_start_time:
            now = datetime.now()
            try:
                slot_dt = datetime.strptime(f"{slot['date']}T{slot_start_time}:00", '%Y-%m-%dT%H:%M:%S')
                diff_ms = (slot_dt - now).total_seconds() * 1000
                if diff_ms < 30 * 60 * 1000:
                    continue
            except ValueError:
                pass

        if slot['otType'] == 'Full Day OT':
            date_entry = next((e for e in agent_entries if e['date'] == slot['date']), None)
            if not date_entry or not date_entry.get('isWeeklyOff'):
                continue
        else:
            if slot.get('assignedAgentId') and slot['assignedAgentId'] != agent_id:
                continue

        result.append(slot)

    return result


def get_manager_programs(manager_name, roster):
    """Returns programs managed by a given manager."""
    programs = set()
    for entry in roster.get('entries', []):
        if entry.get('manager') == manager_name:
            programs.add(entry['program'])
    return list(programs)

File: server/src/modules/accessControl.ts
import type { OTSlot, ShiftRoster, ShiftEntry } from '../types.js';

/**
 * Returns the ISO week string for a date, e.g. "2025-W15".
 */
function getWeek(dateStr: string): string {
  const date = new Date(dateStr + 'T12:00:00Z');
  const dayOfWeek = date.getUTCDay() || 7;
  date.setUTCDate(date.getUTCDate() + 4 - dayOfWeek);
  const yearStart = new Date(Date.UTC(date.getUTCFullYear(), 0, 1));
  const weekNo = Math.ceil(((date.getTime() - yearStart.getTime()) / 86400000 + 1) / 7);
  return `${date.getUTCFullYear()}-W${String(weekNo).padStart(2, '0')}`;
}

/**
 * Validates whether an agent can pick up a specific slot.
 */
export function validateAgentPickup(
  slots: OTSlot[],
  slotId: string,
  agentId: string,
  roster: ShiftRoster,
): { valid: boolean; error?: string } {
  const slot = slots.find((s) => s.id === slotId);
  if (!slot) return { valid: false, error: 'Slot not found.' };

  if (slot.status !== 'Released') {
    return { valid: false, error: 'This slot is no longer available.' };
  }

  // 30-minute cutoff: agent cannot pick up a slot if it starts within 30 minutes
  const slotStartTime = slot.timeWindow.split('-')[0]?.trim();
  if (slotStartTime) {
    const now = new Date();
    const slotDateTime = new Date(`${slot.date}T${slotStartTime}:00`);
    const diffMs = slotDateTime.getTime() - now.getTime();
    const diffMinutes = diffMs / (1000 * 60);
    if (diffMinutes < 30) {
      return { valid: false, error: 'This slot can no longer be picked up (less than 30 minutes to start time).' };
    }
  }

  // Find agent's entries
  const agentEntries = roster.entries.filter((e) => e.agent === agentId);
  if (agentEntries.length === 0) {
    return { valid: false, error: 'Agent not found in roster.' };
  }

  // Check program match
  const agentPrograms = [...new Set(agentEntries.map((e) => e.program))];
  if (!agentPrograms.includes(slot.program)) {
    return { valid: false, error: 'You are not eligible for this slot (program mismatch).' };
  }

  // Check lobby match — agent can only pick up slots from their lobby
  if (slot.lobby) {
    const agentLobbies = [...new Set(agentEntries.filter((e) => e.lobby).map((e) => e.lobby))];
    if (agentLobbies.length > 0 && !agentLobbies.includes(slot.lobby)) {
      return { valid: false, error: 'You are not eligible for this slot (lobby mismatch).' };
    }
  }

  const isFullDay = slot.otType === 'Full Day OT';

  if (isFullDay) {
    // Agent must have weekly off on that date
    const dateEntry = agentEntries.find((e) => e.date === slot.date);
    if (!dateEntry || !dateEntry.isWeeklyOff) {
      return { valid: false, error: 'Full Day OT is only available for agents with a weekly off on this date.' };
    }

    // Max 1 Full Day OT per week when agent has 2+ weekly offs
    const slotWeek = getWeek(slot.date);
    const agentWODates = agentEntries.filter((e) => e.isWeeklyOff).map((e) => e.date);
    const woInWeek = agentWODates.filter((d) => getWeek(d) === slotWeek).length;

    if (woInWeek >= 2) {
      const filledFullDayThisWeek = slots.filter(
        (s) => s.status === 'Filled' &&
          s.filledByAgentId === agentId &&
          s.otType === 'Full Day OT' &&
          getWeek(s.date) === slotWeek
      );
      if (filledFullDayThisWeek.length >= 1) {
        return { valid: false, error: 'You have reached the maximum Full Day OT pickups for this week.' };
      }
    }
  } else {
    // Pre/Post OT: slot's assigned agent must match requesting agent
    if (slot.assignedAgentId && slot.assignedAgentId !== agentId) {
      return { valid: false, error: 'This slot is assigned to a different agent.' };
    }

    // Max 1 Pre/Post OT per agent per day
    const filledPrePostToday = slots.filter(
      (s) => s.status === 'Filled' &&
        s.filledByAgentId === agentId &&
        s.date === slot.date &&
        s.otType !== 'Full Day OT'
    );
    if (filledPrePostToday.length >= 1) {
      return { valid: false, error: 'You already have a Pre/Post Shift OT pickup on this date.' };
    }
  }

  return { valid: true };
}

/**
 * Returns eligible slots for an agent.
 */
export function getEligibleSlotsForAgent(
  slots: OTSlot[],
  agentId: string,
  roster: ShiftRoster,
): OTSlot[] {
  const agentEntries = roster.entries.filter((e) => e.agent === agentId);
  if (agentEntries.length === 0) return [];

  const agentPrograms = new Set(agentEntries.map((e) => e.program));
  const agentLobbies = new Set(agentEntries.filter((e) => e.lobby).map((e) => e.lobby));
  const releasedSlots = slots.filter((s) => s.status === 'Released');

  return releasedSlots.filter((slot) => {
    // Must match program
    if (!agentPrograms.has(slot.program)) return false;

    // Must match lobby (if slot has a lobby and agent has lobbies)
    if (slot.lobby && agentLobbies.size > 0 && !agentLobbies.has(slot.lobby)) return false;

    // 30-minute cutoff: hide slots that start within 30 minutes
    const slotStartTime = slot.timeWindow.split('-')[0]?.trim();
    if (slotStartTime) {
      const now = new Date();
      const slotDateTime = new Date(`${slot.date}T${slotStartTime}:00`);
      const diffMs = slotDateTime.getTime() - now.getTime();
      if (diffMs < 30 * 60 * 1000) return false;
    }

    if (slot.otType === 'Full Day OT') {
      // Agent must have weekly off on that date
      const dateEntry = agentEntries.find((e) => e.date === slot.date);
      return dateEntry?.isWeeklyOff === true;
    } else {
      // Pre/Post OT: assigned agent must match
      if (slot.assignedAgentId && slot.assignedAgentId !== agentId) return false;
      return true;
    }
  });
}

/**
 * Returns programs managed by a given manager.
 */
export function getManagerPrograms(managerName: string, roster: ShiftRoster): string[] {
  const programs = new Set<string>();
  for (const entry of roster.entries) {
    if (entry.manager === managerName) {
      programs.add(entry.program);
    }
  }
  return [...programs];
}
